### PR TITLE
Backend Work for Crypto Wallets reset feature

### PR DESCRIPTION
**Pull request by ryanml**
Partially fixes: https://github.com/brave/brave-browser/issues/5741
Fixes: https://github.com/brave/brave-browser/issues/7716

After the deletion of the DB entry, two things have to happen:

1. All tabs containing `brave://wallet` need to close
2. Brave must restart

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
Note: The Extension at this time lacks the `"management"` permissions allowing programmatic reload, so the above snippet will not work. However, these steps can be used to verify functionality:

1. Clean profile, navigate to `brave://wallet`
2. Import/create a new wallet
3. Confirm wallet has been created.
4. In that page, open up the console.
5. Run:
```
chrome.braveWallet.resetWallet()
```
6. Crypto wallets tabs will close and Brave will restart
7. Navigate to `brave://wallet`, confirm state is fresh
8. Do the above steps then both create a new wallet and import an account and confirm this works as expected. 

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_wallet/browser/brave_wallet_controller.cc [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+      base64_cipher_seed);
+}
+
+void BraveWalletController::ResetCryptoWallets(
+    ResetCryptoWalletsCallback callback) {
+  Profile* profile = Profile::FromBrowserContext(context_);
+  const base::FilePath wallet_data_path = profile->GetPath()
+      .AppendASCII("Local Extension Settings")
+      .AppendASCII(ethereum_remote_client_extension_id)
+      .AppendASCII("000003.LOG");
```
Comment (ryanml) [inline components/brave_wallet/browser/brave_wallet_controller.cc]: It appears in rare cases this filename can differ, better to just remove the directory

Comment (ryanml) [reply components/brave_wallet/browser/brave_wallet_controller.cc]: now just removing dir

File: browser/extensions/api/brave_wallet_api.cc [hunk 2]
```diff
@@ -21,19 +20,27 @@
 #include "chrome/browser/profiles/profile.h"
 #include "components/prefs/pref_service.h"
 #include "content/public/browser/web_contents.h"
-#include "crypto/aead.h"
-#include "crypto/hkdf.h"
-#include "crypto/random.h"
-#include "crypto/symmetric_key.h"
 #include "extensions/browser/extension_registry.h"
 #include "extensions/browser/extension_util.h"
+#include "base/files/file_util.h"
```
Comment (bbondy) [inline browser/extensions/api/brave_wallet_api.cc]: I think this is not used in this file, probably was when you were developing at one point. I think you can remove it.

File: components/brave_wallet/browser/brave_wallet_controller.cc [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+void BraveWalletController::CloseTabsAndRestart() {
+  // Close all CW tabs in each browser instance
+  for (auto* browser : *BrowserList::GetInstance()) {
+    auto* tab_strip = browser->tab_strip_model();
+    for (int i = 0; i < tab_strip->count(); ++i) {
+      auto* web_contents = tab_strip->GetWebContentsAt(i);
+      GURL url = web_contents->GetURL();
+      if (url.SchemeIs(content::kChromeUIScheme) &&
+          url.host() == ethereum_remote_client_host) {
+        web_contents->Close();
```
Comment (bbondy) [inline components/brave_wallet/browser/brave_wallet_controller.cc]: Does this actually get rid of the tab_strip's webcontents too? if so this might cause some wonkiness with the iterator on the count.

Comment (ryanml) [reply components/brave_wallet/browser/brave_wallet_controller.cc]: I'll do some tests and double check, I'll report back 👍

File: browser/extensions/api/brave_wallet_api.cc (hunk 2 above)
Comment (ryanml) [reply browser/extensions/api/brave_wallet_api.cc]: Ah yes, will remove

Comment (ryanml) [reply browser/extensions/api/brave_wallet_api.cc]: removed

File: components/brave_wallet/browser/brave_wallet_controller.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+#include "base/memory/scoped_refptr.h"
+#include "base/memory/weak_ptr.h"
+#include "base/observer_list.h"
+#include "base/values.h"
+#include "url/gurl.h"
+#include "chrome/browser/profiles/profile.h"
+
+namespace base {
+class FilePath;
+class SequencedTaskRunner;
```
Comment (bridiver) [inline components/brave_wallet/browser/brave_wallet_controller.h]: why is this here? It isn't used anywhere

File: components/brave_wallet/browser/brave_wallet_service.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "base/files/file_path.h"
+#include "base/macros.h"
+#include "base/memory/scoped_refptr.h"
+#include "base/memory/weak_ptr.h"
+#include "base/scoped_observer.h"
+#include "components/keyed_service/core/keyed_service.h"
+
+namespace base {
+class SequencedTaskRunner;
```
Comment (bridiver) [inline components/brave_wallet/browser/brave_wallet_service.h]: why is this here? It isn't referenced anywhere in the header

File: components/brave_wallet/browser/brave_wallet_service.cc [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+#include "base/task/post_task.h"
+#include "base/task_runner_util.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_controller.h"
+#include "content/public/browser/browser_context.h"
+
+BraveWalletService::BraveWalletService(content::BrowserContext* context)
+    : context_(context),
+      controller_(new BraveWalletController(context)),
+      weak_factory_(this) {
```
Comment (bridiver) [inline components/brave_wallet/browser/brave_wallet_service.cc]: why does this have a weak ptr? It is never used for anything

File: components/brave_wallet/browser/brave_wallet_service_factory.cc [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+#include "chrome/browser/profiles/profile.h"
+#include "components/keyed_service/content/browser_context_dependency_manager.h"
+
+// static
+BraveWalletServiceFactory* BraveWalletServiceFactory::GetInstance() {
+  return base::Singleton<BraveWalletServiceFactory>::get();
+}
+
+// static
+BraveWalletService* BraveWalletServiceFactory::GetForProfile(Profile* profile) {
```
Comment (bridiver) [inline components/brave_wallet/browser/brave_wallet_service_factory.cc]: this factory needs to go in `brave/browser/brave_wallet`

Comment (bbondy) [reply components/brave_wallet/browser/brave_wallet_service_factory.cc]: I moved that already in a diff PR a few weeks ago

**Comment by urbenlegend** [comment]
Has the fix been released into Brave stable yet? Currently `chrome.braveWallet.resetWallet()` does nothing when run in the console on the Wallets page.

**Comment by paulvi** [comment]
This PR closes https://github.com/brave/brave-browser/issues/5741 however there is no hint how to actually remove/initialize Brave Wallet

@urbenlegend  
> chrome.braveWallet.resetWallet()  run in the console on the Wallets page.
Where did you find this action suggestion?
